feat(docs-platform): Migrate style guide (#5052)
Kong/developer.konghq.com: app/contributing/style-guide.md

---
title: "Style guide"
content_type: reference
layout: reference

breadcrumbs:
  - /contributing/

tags:
  - contributing

description: "Writing and formatting guidelines for contributing to the Kong Developer site. Covers language, tone, grammar, capitalization, code formatting, icons, and third-party tool documentation."

products:
  - gateway

works_on:
  - on-prem
  - konnect

llm: false
---

> 

Author: ```suggestion
Writing and formatting guidelines for contributing to the Kong Developer site. Covers language, tone, grammar, capitalization, code formatting, icons, and third-party tool documentation.
```

<!--vale off-->

## Language

> The Kong docs use [American English (US English)](https://en.wikipedia.org/wiki/American_English). Pay attention to spelling and vocabulary differences, particularly when reviewing contributions from international authors.

Reviewer: ```suggestion
The Kong docs use [American English (US English)](https://en.wikipedia.org/wiki/American_English).
```
I'm not sure that we need this.

      For example: "Once you have added *the inputs section*, ..."
    dont: |
      Avoid generic pronouns.
      For example, don't say: "Once you have added *this*, ..."
{% endtable %}

### Active voice

Use active voice, and avoid passive voice.

With active voice, the subject performs an action. With passive voice, the action is performed upon the subject:

{% table %}
columns:
  - title: "✅ Active"
    key: active
  - title: "❌ Passive"
    key: passive
rows:
  - active: The plugin *applies* rate limiting to consumers.
    passive: Rate limits *are applied to* consumers by the plugin.
  - active: You *can explore* the API using a browser.
    passive: The API *can be explored* using a browser.
  - active: The YAML file *specifies* the replica count.
    passive: The replica count *is specified* in the YAML file.
{% endtable %}

There are exceptions. You might use passive voice when the subject is genuinely passive and isn't performing any action.
For example: "The CA provider's external account can be registered automatically."
In this sentence, the account isn't doing anything, so passive voice is appropriate.

### Present tense for references

> Whenever possible, use present tense instead of past or future tense for reference documentation.

Reviewer: ```suggestion
Whenever possible, use present tense instead of past or future tense for documentation.
```
I think this applies to all doc types, I think if we have reference called out here, it might skip this advice for landing pages/how-tos

{% table %}
columns:
  - title: "✅ Do use"
    key: do
  - title: "❌ Don't use"
    key: dont
rows:
  - do: This `command` *starts* a proxy.
    dont: This `command` *will start* a proxy.
  - do: This `command` *starts* a proxy.
    dont: This `command` *has started* a proxy.
{% endtable %}

Just like for active voice, there are exceptions here too. 
In how-to guides, we use a more natural tone, so using "will" in phrases like "You'll see this result on the screen" is completely fine. 

### Contractions

Don't be afraid to use contractions (*can't*, *isn't*, *you'll*, and so on). They contribute to our conversational tone.

There are exceptions. Omit contractions when aiming for a more serious tone, such as in a warning or caution:

- **Contraction (informational):** "This plugin isn't available in Konnect."

> - **No contraction (warning):** "**Do not** use this plugin in Konnect because it will break things."

Reviewer: ```suggestion
- **No contraction (warning):** "**Do not** use this plugin in Konnect because it will break your configuration."
```
Adjusting wording here so AI doesn't think "break things" is an okay phrasing 🙂

### Latin phrases

Don't use Latin abbreviations or phrases. Use the English equivalent instead.

{% table %}
columns:
  - title: "✅ Do use"
    key: do
  - title: "❌ Don't use"
    key: dont
rows:
  - do: "For example, ..."

>     dont: "e.g., ..."

Reviewer: ```suggestion
    dont: "e.g., ex., ..."
```

- Certificate
- Consumer
- Plugin
- Route
- Service (Gateway Service)
- Target
- Upstream
- Vault

The following terms should not be capitalized. They should be lowercase unless at the start of a sentence:
* control plane
* data plane
* application
* database
* developer
* hybrid mode
* service mesh

#### Plugin name capitalization

1. Capitalize the plugin *name* but not the word *plugin*. For example, "Rate Limiting plugin".
2. Don't capitalize in code. For example, `rate-limiting`.
3. Don't capitalize if you're referring to the concept rather than the plugin itself.
   For example: "Set up rate limiting in {{site.base_gateway}} with the Rate Limiting plugin."

## Formatting

### Placeholder and example values

The type of placeholder you use depends on context:

- **Generic placeholder values** — in most situations (such as plugin parameters, YAML examples, or Kong configuration), use all-caps text with underscores between words.

  For example: `service: SERVICE_NAME`

- **Placeholders in API URLs or OpenAPI specs** — enclose in `{ }` and use the parameter name defined by the API or spec, per [Swagger guidelines](https://swagger.io/docs/specification/describing-parameters/).

>   For example: `/services/{serviceID}/plugins`

Reviewer: ```suggestion
  For example: `/services/{serviceId}/plugins`
```
I think it's this

- **Hostnames and example URLs:**

>   - For guides with examples intended to be run as-is, use `localhost` as the domain.

Reviewer: ```suggestion
  - For guides with examples intended to be run as-is, use `localhost` as the domain or `$KONNECT_PROXY_URL` for {{site.konnect_short_name}} documentation.
```

    For example: `curl -i -X GET https://localhost:8001/services`

    A reader following this guide with {{site.base_gateway}} running locally can copy and paste the command directly into a terminal.

  - For illustrative examples that are not intended to be run as-is, use `example` or `example.com`.

    For example: `user@example.com` or `https://example.okta.admin.com`

- **Path parameters** — always denote with curly braces `{}`.

>   For example: `http://localhost:8001/services/{serviceID|serviceName}/routes/{routeID|routeName}`

Reviewer: ```suggestion
  For example: `http://localhost:8001/services/{serviceId|serviceName}/routes/{routeId|routeName}`
```

#### Inline placeholders

If you're adding a placeholder inline in a sentence, enclose it in single backticks: \`EXAMPLE_TEXT\`

### Code formatting

> - Separate commands from their output. Put each in its own code block.

Reviewer: ```suggestion
- Separate commands from their output. Put each in its own code block. Use `{:.no-copy-code}` directly under an output if users won't need to copy the output.
```

- Include properly formatted code comments.

> - For long commands, split the code block across multiple lines using `\` to avoid horizontal scrolling.

Reviewer: I've never done this (but maybe should have), is this still a valid recommendation?

Author: You have! It's what we do with all our curl commands, if we're writing them manually 😄 definitely still valid

- Never include more than one command in a single code block.

> - Always set a language for code blocks (for example, `bash`, `yaml`) to enable syntax highlighting.

Reviewer: ```suggestion
- Use yaml liquid blocks for code when it's supported (for example, `entity_example`, `konnect_api_request`). If there are no corresponding liquid blocks, always set a language for code blocks (for example, `bash`, `yaml`) to enable syntax highlighting.
```

- Do **not** use the command prompt marker (`$`) in code snippets.

## Icons

For inline icons in prose or tables, use SVG files from the [`/app/assets/icons/`](https://github.com/Kong/developer.konghq.com/tree/main/app/assets/icons) directory. Browse the directory to find the icon you need, then reference it with a relative path.

## Documenting third-party tools

> When a how-to guide requires a third-party tool (such as an identity provider, cloud service, or external API), include complete setup instructions as a prerequisite rather than linking out to third-party documentation and expecting readers to figure it out.

Reviewer: ```suggestion
When a how-to guide requires a third-party tool (such as an identity provider, cloud service, or external API) to be configured in a specific way to work with Kong, include complete setup instructions rather than linking out to third-party documentation and expecting readers to figure it out.
```
Maybe a hot take/not everyone will agree with my rewrite? For example, if I just need an AWS IAM user and the configuration isn't specific to Kong, I'll link out to AWS docs for that.

Author: Good call - I don't think that's a hot take at all. I'm pretty sure we normally do this.

> Write the prerequisite steps in a file under `app/_includes/prereqs/` and include it at the top of the how-to. This keeps the how-to self-contained: readers can follow the entire guide in one place without switching between multiple sources.

Reviewer: ```suggestion
In most cases, the third-party instructions should be a prerequisite.
Write the prerequisite steps in a file under `app/_includes/prereqs/` and include it at the top of the how-to. This keeps the how-to self-contained: readers can follow the entire guide in one place without switching between multiple sources.
```

> For an example, see [`app/_includes/prereqs/gemini.md`](https://github.com/Kong/developer.konghq.com/blob/main/app/_includes/prereqs/gemini.md), which walks users through getting a Gemini API key from Google Cloud before the main tutorial begins.

Reviewer: ```suggestion
For an example, see [`app/_includes/prereqs/gemini.md`](https://github.com/Kong/developer.konghq.com/blob/main/app/_includes/prereqs/gemini.md), which walks users through getting a Gemini API key from Google Cloud before the main tutorial begins.

In certain cases, the third-party instructions should be in the main how-to body.
For example, if users need to update a routing table in AWS after they've set up their Dedicated Cloud Gateway network or if they need to approve a Catalog integration in the third-party provider after they configure it in Catalog.
```
Tried to get at the nuance there because sometimes the third-party steps shouldn't be in the prereqs.
